src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -996,6 +1020,74 @@ pub mod tests {
         assert_eq!(net.acked_features, features);
     }
 
+    #[test]
+    // We can't get offload features, that were set up to a TAP device by ioctl,
+    // hence that we have to validate, that we sort out unsupported features correctly
+    fn test_build_tap_supported_features_all() {
+        let desired_features = [
+            (gen::TUN_F_CSUM, VIRTIO_NET_F_CSUM),
+            (gen::TUN_F_UFO, VIRTIO_NET_F_GUEST_UFO),
+            (gen::TUN_F_TSO4, VIRTIO_NET_F_GUEST_TSO4),
+            (gen::TUN_F_TSO6, VIRTIO_NET_F_GUEST_TSO6),
+        ];
+
+        let mut tap_features: u32 = 0;
+        let mut ack_features: u32 = 0;
+
+        for i in 0..desired_features.len() {
+            tap_features |= desired_features[i].0;
+            ack_features |= 1 << desired_features[i].1;
+        }
+
+        let supported_flags =
+            Net::build_tap_supported_features(&desired_features, ack_features as u64);
+
+        assert_eq!(supported_flags, tap_features);
+    }
+
+    #[test]
+    fn test_build_tap_supported_features_ack_les_than_want() {
+        let desired_features = [
+            (gen::TUN_F_CSUM, VIRTIO_NET_F_CSUM),
+            (gen::TUN_F_UFO, VIRTIO_NET_F_GUEST_UFO),
+            (gen::TUN_F_TSO4, VIRTIO_NET_F_GUEST_TSO4),
+            (gen::TUN_F_TSO6, VIRTIO_NET_F_GUEST_TSO6),
+        ];
+
+        let tap_features: u32 = gen::TUN_F_CSUM | gen::TUN_F_UFO;
+        let ack_features: u32 = 1 << VIRTIO_NET_F_CSUM | 1 << VIRTIO_NET_F_GUEST_UFO;
+
+        let supported_flags =
+            Net::build_tap_supported_features(&desired_features, ack_features as u64);
+
+        assert_eq!(supported_flags, tap_features);
+    }
+
+    #[test]
+    fn test_build_tap_supported_features_ack_more_than_want() {
+        let desired_features = [
+            (gen::TUN_F_CSUM, VIRTIO_NET_F_CSUM),
+            (gen::TUN_F_UFO, VIRTIO_NET_F_GUEST_UFO),
+            (gen::TUN_F_TSO4, VIRTIO_NET_F_GUEST_TSO4),
+        ];
+
+        let mut tap_features: u32 = 0;
+        let mut ack_features: u32 = 0;
+
+        for i in 0..desired_features.len() {
+            tap_features |= desired_features[i].0;
+            ack_features |= 1 << desired_features[i].1;
+        }
+
+        // extra supported feature
+        ack_features |= 1 << VIRTIO_NET_F_GUEST_TSO6;
```

Comment:
Good that you added this version but I think this cannot happen. The MMIO layer will make sure that we will never see more flags than asked for in the acknowledged features.

Reply:
I my be wrong, but it sounds to me like we rely on 3rd party thing to prevent such behaviour. Since we add flags here and now it looks reasonable to me to test it in the same place.

But I have no idea about FC and maybe virtio device and MMIO layer are inseparable things indeed.

Reply:
MMIO layer is implemented by Firecracker. Here: https://github.com/firecracker-microvm/firecracker/blob/main/src/vmm/src/devices/virtio/device.rs#L155 is where we check that we will not accept acked features that we did not offer.

Reply:
Oh, sorry, I didn't mean it's 3rd party lib. I was on about relying on logic in a different place. And if it's changed once we may loose this coverage. Or we should leave a comment that we need this check there, cos we expect it in the different place.

It's more about global pattern, rather than this case in particular. If you are ok with that - I can delete this test.